File: Examples/FriendlySiteMapProvider/trunk/PinballWizard_CS/App_Code/PinballSiteMapProvider.cs
using System;
using System.Data;
using System.Configuration;
using System.Web;
using System.Collections.Generic;

/// <summary>
/// Summary description for PinballSiteMapProvider
/// </summary>
public class PinballSiteMapProvider : SiteMapProvider
{
	private SiteMapNode _smnRoot;
	private SiteMapNodeCollection _smncCategory;
	private Dictionary<string, SiteMapNodeCollection> _dArticle;
	private object lockobj = new object();

	public void BuildNodes() {
		lock (lockobj) {
			_smnRoot = new SiteMapNode(this, "PinballRoot", "~/Default.aspx", "Pinball Wizard");
			_smncCategory = new SiteMapNodeCollection();
			_dArticle = new Dictionary<string, SiteMapNodeCollection>();

			ContentTableAdapters.ArticleTableAdapter taArticle = new ContentTableAdapters.ArticleTableAdapter();
			Content.ArticleDataTable dtArticle = taArticle.AllArticles();

			foreach (Content.ArticleRow row in dtArticle.Rows) {
				SiteMapNode smnNew = new SiteMapNode(this, row.article_id.ToString(), MakeURL(row.title, row.category, row.post_date), row.title);

				if (!_dArticle.ContainsKey(row.code)) {
					SiteMapNode smnNewCat = new SiteMapNode(this, row.code, MakeURL(row.category), row.category);
					smnNew.ParentNode = smnNewCat;
					_smncCategory.Add(smnNewCat);
					_dArticle[row.code] = new SiteMapNodeCollection();
				} else {
					smnNew.ParentNode = _dArticle[row.code][0].ParentNode;
				}
				
				_dArticle[row.code].Add(smnNew);
			}
		}
	}

	public static String Sanitize(String data) {
		String allowed = "abcdefghijklmnopqrstuvwxyz0123456789.";
		String clean = "";

		foreach (Char l in data.ToCharArray()) {
			if (allowed.Contains(l.ToString().ToLower())) {
				clean += l;
			} else if (l.Equals(' ') && (clean.Length > 0 && clean[clean.Length - 1] != '-')) {
				clean += "-";
			}
		}

		return clean;
	}

	public static String MakeURL(String title, String category, DateTime post_date) {
		String url = "";
		title = Sanitize(title);
		category = Sanitize(category);

		url = "~/Articles/" + category + post_date.ToString("/yyy/MM/dd/") + title + ".aspx";

		return url;
	}

	public static String MakeURL(String category) {
		String url = "";
		category = Sanitize(category);

		url = "~/Articles/" + category + "/Default.aspx";

		return url;
	}

	public override void Initialize(string name, System.Collections.Specialized.NameValueCollection attributes) {
		base.Initialize(name, attributes);
		this.BuildNodes();
	}
	
	public override SiteMapNode FindSiteMapNode(string rawUrl) {
		String relUrl = rawUrl.Replace(HttpRuntime.AppDomainAppVirtualPath, "~").ToLower();
		
		if (_smnRoot.Url.ToLower() == relUrl || _smnRoot.Key == rawUrl)
			return _smnRoot;

		foreach (SiteMapNode smnCat in _smncCategory)
			if (smnCat.Url.ToLower() == relUrl || smnCat.Key == rawUrl)
				return smnCat;

		foreach (SiteMapNodeCollection smncArticle in _dArticle.Values)
			foreach (SiteMapNode smnArticle in smncArticle)
				if (smnArticle.Url.ToLower() == relUrl || smnArticle.Key == rawUrl)
					return smnArticle;

		return null;
	}

	public override SiteMapNodeCollection GetChildNodes(SiteMapNode node) {
		if (_smnRoot.Key == node.Key)
			return _smncCategory;

		if (_dArticle.ContainsKey(node.Key))
			return _dArticle[node.Key];

		return null;
	}

	public override SiteMapNode GetParentNode(SiteMapNode node) {
		if (_smnRoot.Key == node.Key)
			return null;

		if (_dArticle.ContainsKey(node.Key))
			return _smnRoot;

		foreach (SiteMapNodeCollection smncArticle in _dArticle.Values)
			foreach (SiteMapNode smnArticle in smncArticle)
				if (smnArticle.Key == node.Key)
					return smnArticle.ParentNode;

		return null;		
	}

	protected override SiteMapNode GetRootNodeCore() {
		return _smnRoot;
	}
}

File: Examples/FriendlySiteMapProvider/trunk/PinballWizard_VB/App_Code/PinballSiteMapProvider.vb
Imports Microsoft.VisualBasic


 _
Public Class PinballSiteMapProvider
	Inherits SiteMapProvider

	Private _smnRoot As SiteMapNode
	Private _smncCategory As SiteMapNodeCollection
	Private _dArticle As Hashtable
	Private lockobj As New Object()


	Public Sub BuildNodes()
		SyncLock lockobj
			_smnRoot = New SiteMapNode(Me, "PinballRoot", "~/Default.aspx", "Pinball Wizard")
			_smncCategory = New SiteMapNodeCollection()
			_dArticle = New Hashtable

			Dim taArticle As New ContentTableAdapters.ArticleTableAdapter()
			Dim dtArticle As Content.ArticleDataTable = taArticle.AllArticles()

			Dim row As Content.ArticleRow
			For Each row In dtArticle.Rows
				Dim smnNew As New SiteMapNode(Me, row.article_id.ToString(), MakeURL(row.title, row.category, row.post_date), row.title)

				If Not _dArticle.ContainsKey(row.code) Then
					Dim smnNewCat As New SiteMapNode(Me, row.code, MakeURL(row.category), row.category)
					smnNew.ParentNode = smnNewCat
					_smncCategory.Add(smnNewCat)
					_dArticle(row.code) = New SiteMapNodeCollection()
				Else
					smnNew.ParentNode = _dArticle(row.code)(0).ParentNode
				End If

				_dArticle(row.code).Add(smnNew)
			Next row
		End SyncLock
	End Sub	'BuildNodes


	Public Shared Function Sanitize(ByVal data As [String]) As [String]
		Dim allowed As [String] = "abcdefghijklmnopqrstuvwxyz0123456789."
		Dim clean As [String] = ""

		Dim l As [Char]
		For Each l In data.ToCharArray()
			If allowed.Contains(l.ToString().ToLower()) Then
				clean += l
			Else
				If l.Equals(" "c) And (clean.Length > 0 And clean((clean.Length - 1)) <> "-"c) Then
					clean += "-"
				End If
			End If
		Next l
		Return clean
	End Function 'Sanitize


	Public Overloads Shared Function MakeURL(ByVal title As [String], ByVal category As [String], ByVal post_date As DateTime) As [String]
		Dim url As [String] = ""
		title = Sanitize(title)
		category = Sanitize(category)

		url = "~/Articles/" + category + post_date.ToString("/yyy/MM/dd/") + title + ".aspx"

		Return url
	End Function 'MakeURL


	Public Overloads Shared Function MakeURL(ByVal category As [String]) As [String]
		Dim url As [String] = ""
		category = Sanitize(category)

		url = "~/Articles/" + category + "/Default.aspx"

		Return url
	End Function 'MakeURL


	Public Overrides Sub Initialize(ByVal name As String, ByVal attributes As System.Collections.Specialized.NameValueCollection)
		MyBase.Initialize(name, attributes)
		Me.BuildNodes()
	End Sub	'Initialize


	Public Overrides Function FindSiteMapNode(ByVal rawUrl As String) As SiteMapNode
		Dim relUrl As [String] = rawUrl.Replace(HttpRuntime.AppDomainAppVirtualPath, "~").ToLower()

		If _smnRoot.Url.ToLower() = relUrl Or _smnRoot.Key = rawUrl Then
			Return _smnRoot
		End If
		Dim smnCat As SiteMapNode
		For Each smnCat In _smncCategory
			If smnCat.Url.ToLower() = relUrl Or smnCat.Key = rawUrl Then
				Return smnCat
			End If
		Next smnCat
		Dim smncArticle As SiteMapNodeCollection
		For Each smncArticle In _dArticle.Values
			Dim smnArticle As SiteMapNode
			For Each smnArticle In smncArticle
				If smnArticle.Url.ToLower() = relUrl Or smnArticle.Key = rawUrl Then
					Return smnArticle
				End If
			Next smnArticle
		Next smncArticle
		Return Nothing
	End Function 'FindSiteMapNode


	Public Overrides Function GetChildNodes(ByVal node As SiteMapNode) As SiteMapNodeCollection
		If _smnRoot.Key = node.Key Then
			Return _smncCategory
		End If
		If _dArticle.ContainsKey(node.Key) Then
			Return _dArticle(node.Key)
		End If
		Return Nothing
	End Function 'GetChildNodes


	Public Overrides Function GetParentNode(ByVal node As SiteMapNode) As SiteMapNode
		If _smnRoot.Key = node.Key Then
			Return Nothing
		End If
		If _dArticle.ContainsKey(node.Key) Then
			Return _smnRoot
		End If
		Dim smncArticle As SiteMapNodeCollection
		For Each smncArticle In _dArticle.Values
			Dim smnArticle As SiteMapNode
			For Each smnArticle In smncArticle
				If smnArticle.Key = node.Key Then
					Return smnArticle.ParentNode
				End If
			Next smnArticle
		Next smncArticle
		Return Nothing
	End Function 'GetParentNode


	Protected Overrides Function GetRootNodeCore() As SiteMapNode
		Return _smnRoot
	End Function 'GetRootNodeCore
End Class 'PinballSiteMapProvider